### ISSUE TITLE
Very wrong positions of minor planets

### ISSUE THREAD

**Issue by alex-w** [Stellarium/stellarium#453]
_Original report by Torsten Merkel:_ https://bugs.launchpad.net/stellarium/+bug/1291005

I tried to observe three of the minor planets during last two weeks (Pallas, Vesta, Ceres) and had problems to find their positions at the night sky using Stellarium 0.12.4 version. With Pallas and Vesta I succeeded at last, but not really satisfied when comparing real constellation with constellation shown in Stellarium. But with Ceres I failed completely last night. At the position where she should have to be seen very clearly there was nothing than black space. Comparing your calculated ephemerides with an annual guide I am using since a couple of years with great success, I realized that Ceres was shown at a position at least

5 DAYS TOO LATE !

Similarly did that happen to Pallas and Vesta positions, but they differred 'only' about two days in movement (too late also), and
Pallas' orbit line is not really what you will observe in the sky, it moves a bit more to the right (smaller values of rectascension I estimate, maybe smaller values of declination too).

Of course this experience was some sort of frustrating for me as I trusted in calculated positions of such well-known objects in the sky even if Stellarium is an amateur developped program with no cost. Now I am irritated a bit whether or
not there is to be trusted in the shown positions of other minor objects.

Maybe there is an unsufficient input of data for ephemerides of minor planets in summary (no disturbing calculation?
wrong equinox or without adjustment to nowadays?). I only can guess, but I dont know. It would be fine if you find
the reason for this problem and it can be solved.

Thank you for listening to my description!

Best regards,

Torsten Merkel

**Comment by gzotti** [comment]
To repeat the final message:
```
This is not a bug at all. Osculating orbital elements for minor planets are 
valid for a rather short time only, as every observer should know. 
This is the reason for having a feature (the solar system editor plugin) to 
update orbital elements, just like many other desktop planetaria have this feature. 
I am not aware of analytic models for orbits of asteroids similar to VSOP87, 
if somebody can provide it, we may consider adding it.

The online resources usually give "current" elements for observing here and now. 
I give the question back: Where could we download past elements? 
How could we date historical photographs of asteroids with software if 
we don't have historical elements?

The only thing we can change may be providing current elements with every new release. 
This should go into the release checklist.
```

**Comment by alex-w** [comment]
Yes, and possible solution for modern days is... automatic updating elements of orbits thorugh internet.

**Comment by gzotti** [comment]
Whoever is going to solve this: Probably "planet objects" should be allowed to have an arbitrary number of osculating orbital elements, with epoch (JDE of perfect validity) added, and computing first has to interpolate from neighboring element sets. For now, just get elements close to the epoch of observation, from wherever you can get them.

**Comment by axd1967** [comment]
Another tool that could be used in validating the results are the ephemerides from JPL's [HORIZONS](https://ssd.jpl.nasa.gov/?horizons).

**Comment by Atque** [comment]
_Hallo northern sky_ uses something called "numerical integration" to get very accurate asteroid positions in the past and future. Perhaps the same procedure could be used in Stellarium?

I have tried this and asteroids have useful positions at least 100 years back (probably even more, if minute of arc errors are acceptable).

https://www.hnsky.org/numint.htm

**Comment by gzotti** [comment]
Sure. Just implement it and send us a pull request :-)
There are a few other issues to solve before one of us may attack this. https://github.com/Stellarium/stellarium/wiki/FAQ#why-dont-you-implement

**Comment by Atque** [comment]
You guys are more knowledgeable than I am, but this seems to be asteroid elements with DE430 ephemeris:

ftp://ssd.jpl.nasa.gov/pub/eph/small_bodies/asteroids_de430/

**Comment by axd1967** [comment]
I've compared the data in 0.20.4 with [HORIZON](https://ssd.jpl.nasa.gov/sbdb.cgi?sstr=ceres;orb=0;cov=0;log=0;cad=0#elem):

![image](https://user-images.githubusercontent.com/3529789/103533657-cbb62800-4e8d-11eb-96d6-91283ef50e07.png)

maybe this seems to explain the differences?

there is also https://www.minorplanetcenter.net/db_search/show_object?object_id=1; but apparently, there is no equivalent data URL for these bodies so they cannot be imported?

not finding Ceres is worrying, but it is known that Stellarium does not yet offer the highest possible precision.

Maybe a first step is to add the "freshness" (epoch/time of last update) of the selected object?

A workaround is to edit `ssystem_minor.ini`? I tried it (using https://www.minorplanetcenter.net/db_search/show_object?object_id=1), and it does not give much difference (but it is possible that I did not update all parameters correctly)

![image](https://user-images.githubusercontent.com/3529789/103535335-ce664c80-4e90-11eb-8c5b-1713b91f9b5b.png)

![image](https://user-images.githubusercontent.com/3529789/103535129-7fb8b280-4e90-11eb-9001-ce4effccf0d4.png)

with my changes, there is not much difference (2'), but maybe I did not update all elements correctly.

**Comment by axd1967** [comment]
( but I just noticed that the OP used 0.12.4 ! )

**Comment by gzotti** [comment]
Make sure orbit_epoch in the element set is close enough to your date of observation.

**Comment by Atque** [comment]
Could SPICE help us here?

https://www.cosmos.esa.int/web/spice

**Comment by gzotti** [comment]
> Could SPICE help us here?
> 
> https://www.cosmos.esa.int/web/spice

If you want to implement it, sure, go ahead.
